### PR TITLE
Add node component and yaml-configured bridge node

### PR DESCRIPTION
**Pull request by mjcarroll**
# 🎉 New feature

## Summary

Primarily introducing a yaml-configured node, but some more refactoring to clean things up.

1. Adds a [component](https://docs.ros2.org/latest/api/rclcpp_components/) version of bridge called `ros_ign_bridge::RosIgnBridge`
  * This component can be used via the stand-alone executable `./bridge_node` or as part of a component container.
  *  This follows best practices for ROS2 going forward.
2. Adds a `yaml`-based configuration that exposes all features of the bridge.  Passed in via a `config_file` parameter.  This should allow us to bypass some of the ERB-and-remapping that we have done on various projects,..

Example yaml configuration for 5 topics below
```
 # Set just topic name, applies to both
- topic_name: "chatter" 
  ros_type_name: "std_msgs/msg/String"
  ign_type_name: "ignition.msgs.StringMsg"

# Set just ROS topic name, applies to both
- ros_topic_name: "chatter_ros" 
  ros_type_name: "std_msgs/msg/String"
  ign_type_name: "ignition.msgs.StringMsg"

# Set just IGN topic name, applies to both
- ign_topic_name: "chatter_ign" 
  ros_type_name: "std_msgs/msg/String"
  ign_type_name: "ignition.msgs.StringMsg"

# Set each topic name explicitly
- ros_topic_name: "chatter_both_ros"
  ign_topic_name: "chatter_both_ign" 
  ros_type_name: "std_msgs/msg/String"
  ign_type_name: "ignition.msgs.StringMsg"

# Full set of configurations
- ros_topic_name: "ros_chatter"
  ign_topic_name: "ign_chatter"
  ros_type_name: "std_msgs/msg/String"
  ign_type_name: "ignition.msgs.StringMsg"
  subscriber_queue: 5
  publisher_queue: 6
  lazy: true
```

3. Refactored `parameter_bridge` and `static_bridge` to use the component.
4. Put `ros_ign_bridge::RosIgnBridge` in the public API so it can be used downstream.

## Test it

Unit tests have been added for the new configuration functionality, no other tests should regress.

Test the new configuration with:
```
./bridge_node --ros-args -p config_file:=$WORKSPACE/ros_ign/ros_ign_bridge/test/config/full.yaml 
```

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: ros_ign_bridge/CMakeLists.txt [hunk 1]
```diff
@@ -52,65 +55,48 @@ set(PACKAGES
...
-    PROPERTIES POSITION_INDEPENDENT_CODE ON)
+  list(APPEND ${bridge_lib}_SRCS src/convert/${PKG}.cpp)
+  list(APPEND ${bridge_lib}_SRCS src/factories/${PKG}.cpp)
 
 endforeach()
 
-set(bridge_executables
-  parameter_bridge
-  static_bridge
-  lazy_static_bridge
```
Comment (iche033) [inline ros_ign_bridge/CMakeLists.txt]: I think its fine to remove this executable if we have not made a release with this yet.

File: ros_ign_bridge/src/bridge_handle.hpp [hunk 2]
```diff
@@ -48,19 +44,13 @@ class Bridge
   /// \param[in] subscriber_queue_size Depth of the subscriber queue
   /// \param[in] publisher_queue_size Depth of the publisher queue
   /// \param[in] is_lazy True for "lazy" subscriptions. (Default: false)
```
Comment (iche033) [inline ros_ign_bridge/src/bridge_handle.hpp]: update doxygen param

File: ros_ign_bridge/src/ros_ign_bridge.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+namespace ros_ign_bridge
+{
+
+RosIgnBridge::RosIgnBridge(const rclcpp::NodeOptions & options)
+: rclcpp::Node("ros_ign_bridge", options)
+{
+  ign_node_ = std::make_shared<ignition::transport::Node>();
+
+  this->declare_parameter<int>("subscription_heartbeat", 1000);
```
Comment (iche033) [inline ros_ign_bridge/src/ros_ign_bridge.cpp:31]: this parameter looks like something new? Document this in `README.md`?

**Review by iche033 — COMMENTED** [review]
There are some conflicts with `galactic` branch

The `Gpr__ros_ign__ubuntu_focal_amd64`  build complains about missing `rclcpp_components` package so I think that extra dependency needs to be added to CI.

yaml based configuration is great new feature. Maybe add some info to `README.md`?

File: ros_ign_bridge/CMakeLists.txt [hunk 4]
```diff
@@ -11,9 +11,12 @@ if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
 endif()
 
 find_package(ament_cmake REQUIRED)
+find_package(rclcpp REQUIRED)
+find_package(rclcpp_components REQUIRED)
+find_package(yaml_cpp_vendor REQUIRED)
+
 find_package(geometry_msgs REQUIRED)
 find_package(nav_msgs REQUIRED)
-find_package(rclcpp REQUIRED)
```
Comment (ahcorde) [inline ros_ign_bridge/CMakeLists.txt]: why do you split this in two sections ? alphabetize ?

File: ros_ign_bridge/README.md [hunk 5]
```diff
@@ -181,3 +181,57 @@ And on terminal E, publish a ROS 2 message:
 `ros2 topic pub /chatter std_msgs/msg/String 'data: "Hello"' -1`
 
 You should see the Ignition listener echoing the message.
+
+## Example 4: Configuring the Bridge via YAML
+
+When configuring many topics, it is easier to use a file-based configuration in a markup
+language. In this case, the ros_ign bridge supports using a YAML file to configure the
```
Comment (ahcorde) [inline ros_ign_bridge/README.md]:
```suggestion
language. In this case, the `ros_ign` bridge supports using a YAML file to configure the
```

File: ros_ign_bridge/README.md [hunk 6]
```diff
@@ -181,3 +181,57 @@ And on terminal E, publish a ROS 2 message:
...
+various parameters.
+
+The configuration file must be a YAML array of maps.
+An example configuration for 5 bridges is below, showing the various ways that a 
+bridge may be specified:
+```
```
Comment (ahcorde) [inline ros_ign_bridge/README.md]:
```suggestion
```yaml
```

File: ros_ign_bridge/README.md [hunk 7]
```diff
@@ -181,3 +181,57 @@ And on terminal E, publish a ROS 2 message:
...
+  ign_topic_name: "ign_chatter"
+  ros_type_name: "std_msgs/msg/String"
+  ign_type_name: "ignition.msgs.StringMsg"
+  subscriber_queue: 5
+  publisher_queue: 6
+  lazy: true
+```
+
+To run the bridge node with the above configuration:
+```
```
Comment (ahcorde) [inline ros_ign_bridge/README.md]:
```suggestion
```bash
```

File: ros_ign_bridge/include/ros_ign_bridge/bridge_config.hpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+/// \param[in] data string containing YAML of bridge configurations
+/// \return Vector of bridge configurations
+std::vector<BridgeConfig> readFromYamlString(const std::string & data);
+
+/// \brief Generate a group of BridgeConfigs from a YAML File
+/// \param[in] filename name of file containing YAML of bridge configurations
+/// \return Vector of bridge configurations
+std::vector<BridgeConfig> readFromYamlFile(const std::string & filename);
+
+
```
Comment (ahcorde) [inline ros_ign_bridge/include/ros_ign_bridge/bridge_config.hpp]:
```suggestion
```

File: ros_ign_bridge/src/bridge_config.cpp [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+  std::ifstream in(filename);
+  return readFromYaml(in);
+}
+
+std::vector<BridgeConfig> readFromYamlString(const std::string & data)
+{
+  std::stringstream ss(data);
+  return readFromYaml(ss);
+}
+
```
Comment (ahcorde) [inline ros_ign_bridge/src/bridge_config.cpp]:
```suggestion
```

File: ros_ign_bridge/src/ros_ign_bridge.cpp [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+  this->declare_parameter<std::string>("config_file", "");
+
```
Comment (ahcorde) [inline ros_ign_bridge/src/ros_ign_bridge.cpp:33]:
```suggestion
```

File: ros_ign_bridge/src/ros_ign_bridge.cpp [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+      this->get_logger(),
+      "Failed to create a bridge for topic [%s] with ROS2 type [%s] "
+      "to topic [%s] with Ignition Transport type [%s]",
+      config.ros_topic_name.c_str(),
+      config.ign_topic_name.c_str(),
+      config.ros_type_name.c_str(),
+      config.ign_type_name.c_str());
+  }
+}
+
```
Comment (ahcorde) [inline ros_ign_bridge/src/ros_ign_bridge.cpp]:
```suggestion
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Review by ahcorde — APPROVED** [review]

**Pull request merged**